psutil/arch/linux/users.c

```diff
@@ -5,13 +5,134 @@
  */
 
 #include <Python.h>
+#include <dlfcn.h>
 #include <utmp.h>
 
 #include "../../_psutil_common.h"
 
+/* Systemd function signatures that will be loaded */
+int (*sd_booted)(void);
+int (*sd_get_sessions)(char ***);
+int (*sd_session_get_leader)(const char *, pid_t *);
+int (*sd_session_get_remote_host)(const char *,char **);
+int (*sd_session_get_start_time)(const char *, uint64_t *);
+int (*sd_session_get_tty)(const char *, char **);
+int (*sd_session_get_username)(const char *, char **);
 
-PyObject *
-psutil_users(PyObject *self, PyObject *args) {
+#define dlsym_check(__h, __fn) do {           \
+    __fn = dlsym(__h, #__fn);                 \
+    if (dlerror() != NULL || __fn == NULL) {  \
+        dlclose(__h);                         \
+        return NULL;                          \
+    }                                         \
+} while (0)
```

Comment:
Is because of the "return". A macro can exit the load_systemd funcion, so I can avoid "if (!dlsym_check()) { return NULL; }"

Reply:
Understood. I think it makes more sense for this to be a function, since it makes things more explicit (e.g. I didn't remember that macros worked like this). You can make things shorter without using parenthesis:

```c
    if (! dlsym_check(handle, sd_booted))
         return NULL;
 ```
 
 Also, it would be good to add some debug info in case a function does not exist. I tried your code, which failed because one function was missing. I had to add this to discover that it was `sd_session_get_leader`:
 
 ```diff
- #define dlsym_check(__h, __fn) do {           \
+ #define dlsym_check(__h, __fn, __name) do {           \
    __fn = dlsym(__h, #__fn);                 \
    if (dlerror() != NULL || __fn == NULL) {  \
+      psutil_debug("missing '%s' fun", __name); \
        dlclose(__h);                         \
        return NULL;                          \
    }                                         \
} while (0)
```